### PR TITLE
first commit

### PR DESCRIPTION
**Pull request by Hester1937**
inclass practise

### CODE REVIEW

File: src/main/java/StringManipulation.java [hunk 1]
```diff
@@ -1,33 +1,74 @@
...
 
     @Override
     public String[] getSubStrings(int startWord, int endWord) {
-        return null;
+        if (string == null || string.isEmpty()) {
+            return new String[0];
+        }
+        String[] words = string.trim().split("\\s+");
+        if (startWord < 0 || startWord >= words.length || endWord < 0 || endWord >= words.length || startWord > endWord) {
+            return new String[0];
```
Comment (jonathanjarman) [inline src/main/java/StringManipulation.java]: You could also throw an indexOutofBounds exception here to help with troubleshooting. But returning an empty array is also fine.

File: src/main/java/StringManipulation.java [hunk 2]
```diff
@@ -1,33 +1,74 @@
...
 
     @Override
     public String restoreString(int[] indices) {
-        return null;
+        if (string == null || string.isEmpty() || indices == null || indices.length != string.length()) {
+            return "";
+        }
+        char[] chars = string.toCharArray();
+        char[] restored = new char[indices.length];
+        for (int i = 0; i < indices.length; i++) {
```
Comment (jonathanjarman) [inline src/main/java/StringManipulation.java:108]: You could also add a check here to ensure the indicies are not smaller than 0 or greater than chars.length.

File: src/test/java/StringManipulationTest.java [hunk 3]
```diff
@@ -28,17 +28,29 @@ public void testCount1() {
...
+        manipulatedstring.setString("  Spaces   between  words ");
+        int length = manipulatedstring.count();
+        assertEquals(3, length);
+    }
+
+    public void testCount5() {
+        manipulatedstring.setString(" ABCDEFD ");
+        int length = manipulatedstring.count();
+        assertEquals(1, length);
     }
```
Comment (jonathanjarman) [inline src/test/java/StringManipulationTest.java:56]: I would add another test method using assertThrows to test throwing exceptions for this method. Such as if you set the string to null you can then throw a null Pointer exception for better testing purposes.

File: src/test/java/StringManipulationTest.java [hunk 4]
```diff
@@ -55,27 +67,37 @@ public void testRemoveNthCharacter2() {
...
+        assertEquals("Hello, Wold!", result);
     }
 
     @Test
     public void testRemoveNthCharacter7() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("Testing");
+        String result = manipulatedstring.removeNthCharacter(10, false);
+        assertEquals("Testing", result);
     }
```
Comment (jonathanjarman) [inline src/test/java/StringManipulationTest.java:109]: Another test case you could write would to have a negative number for N (such as manipulatedstring.setString(-1,false);
You could then assertThrows to throw an exception based on the feedback I provided in the StringManipulation file. This is just another edge case that could be encountered.

File: src/test/java/StringManipulationTest.java [hunk 5]
```diff
@@ -88,24 +110,41 @@ public void testGeSubStrings1() {
...
-    }
-    @Test
-    public void testGeSubStrings6() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("Java is a programming language");
+        String[] subStrings = manipulatedstring.getSubStrings(4, 6);
+
+        assertEquals("programming", subStrings[0]);
+        assertEquals("language", subStrings[1]);
     }
```
Comment (jonathanjarman) [inline src/test/java/StringManipulationTest.java:154]: Another test case you could write would be to have the substring you are trying to retrieve be larger than the length of words in the string. Such as manipulatedstring.getSubstring(122,2000); Then use assertThrows to throw an indexOutOfBounds Exception.

**Review by jonathanjarman — APPROVED** [review]
Overall This was very good! One recommendation I noticed throughout the methods is that there are no exceptions being thrown and then getting tested in the testing class. This can be useful for out of bounds exceptions or cases where an input is null, although returning zero is also acceptable for null cases. Throwing exceptions can be useful for if an error is encountered during critical operations and can help with error handling.

**Review by jonathanjarman — CHANGES_REQUESTED** [review]
Try removing the Main.java file and then checking Travis CI to see if the build is able to pass.

**Review by jonathanjarman — CHANGES_REQUESTED** [review]
In this latest build, you have an extra folder and file. This may be why the build is failing.  org/example should be deleted. Then try pushing the commit to see if the build passes. 
image

File: src/main/java/StringManipulation.java [hunk 6]
```diff
@@ -35,7 +35,7 @@ public int count() {
 
     @Override
     public String removeNthCharacter(int n, boolean maintainSpacing) {
-        
+
         if (this.string == null || this.string.isEmpty()){
```
Comment (jonathanjarman) [inline src/main/java/StringManipulation.java:39]: In this if statement (string == null) and string.isEmpty() are not the same. You can throw a null pointer exception in the case of a null pointer, however it would be better to have another if statement to check if the string is empty, such as string = "".  The string.isEmpty() will be true because it will contain no characters. In this case you can just return 0.

File: src/test/java/StringManipulationTest.java [hunk 7]
```diff
@@ -36,8 +36,8 @@ public void testCount2() {
 
     @Test
     public void testCount3() {
-        assertThrows(IllegalArgumentException.class, () -> { 
-        int length = manipulatedstring.count();
+        assertThrows(NullPointerException.class, () -> {
+            int length = manipulatedstring.count();
```
Comment (jonathanjarman) [inline src/test/java/StringManipulationTest.java]: In this method int length is not needed as it won't be used. The second condition for the string is achieve from manipulatedstring.count(). you can remove "int length =" as it is not necessary.

File: src/test/java/StringManipulationTest.java [hunk 8]
```diff
@@ -120,7 +122,7 @@ public void testGetSubStrings2() {
         assertThrows(IllegalArgumentException.class, () -> {
             manipulatedstring.setString("OnlyOneWord");
             String[] subStrings = manipulatedstring.getSubStrings(1, -1);
```
Comment (jonathanjarman) [inline src/test/java/StringManipulationTest.java]: In this method String[] subStrings is not needed.  You can just delete it and leave the manipulatedstring part after it. Photo for reference. 
image

File: src/test/java/StringManipulationTest.java [hunk 9]
```diff
@@ -167,33 +169,33 @@ public void testRestoreString2() {
             manipulatedstring.setString("testing");
             int[] indices = {};
             String restoredString = manipulatedstring.restoreString(indices);
```
Comment (jonathanjarman) [inline src/test/java/StringManipulationTest.java]: You can remove the local variable String restoredString as it is not needed since the error will be thrown by the assertThrows. Picture for reference. 
image

File: src/test/java/StringManipulationTest.java [hunk 10]
```diff
@@ -167,33 +169,33 @@ public void testRestoreString2() {
...
+        });
     }
 
     @Test
     public void testRestoreString3() {
-        assertThrows(NullPointerException.class,() ->{ 
-            manipulatedstring.setString("");
+        assertThrows(NullPointerException.class,() ->{
             int[] indices = {};
             String restoredString = manipulatedstring.restoreString(indices);
```
Comment (jonathanjarman) [inline src/test/java/StringManipulationTest.java]: You can remove the local variable String restoredString as it is not needed since the error will be thrown by the assertThrows. Picture for reference. 
image

File: src/test/java/StringManipulationTest.java [hunk 11]
```diff
@@ -167,33 +169,33 @@ public void testRestoreString2() {
...
+        });
     }
 
-   @Test
+    @Test
     public void testRestoreString4() {
         assertThrows(IndexOutOfBoundsException.class, () -> {
             manipulatedstring.setString("hello");
             int[] indices = {4, 3, 2, 1, -1};
             String restoredString = manipulatedstring.restoreString(indices);
```
Comment (jonathanjarman) [inline src/test/java/StringManipulationTest.java]: You can remove the local variable String restoredString as it is not needed since the error will be thrown by the assertThrows. Picture for reference.
image

File: src/test/java/StringManipulationTest.java [hunk 12]
```diff
@@ -167,33 +169,33 @@ public void testRestoreString2() {
...
     @Test
     public void testRestoreString5() {
-        manipulatedstring.setString("world");
-        int[] indices = {0, 2, 1, 3, 5};
-        String restoredString = manipulatedstring.restoreString(indices);
-        assertEquals("", restoredString);
+        assertThrows(IndexOutOfBoundsException.class, () -> {
+            manipulatedstring.setString("world");
+            int[] indices = {0, 2, 1, 3, 5};
+            String restoredString = manipulatedstring.restoreString(indices);
```
Comment (jonathanjarman) [inline src/test/java/StringManipulationTest.java]: You can remove the local variable String restoredString as it is not needed since the error will be thrown by the assertThrows. Picture for reference.
image

File: src/main/java/StringManipulation.java [hunk 13]
```diff
@@ -14,61 +14,86 @@ public void setString(String string) {
...
         }
-        int numSubstrings = endWord - startWord + 1;
-        String[] substrings = new String[numSubstrings];
+
+        String[] subStrings = new String[endWord - startWord + 1];
         int index = 0;
-        for (int i = startWord; i <= endWord; i++) {
-            substrings[index] = words[i];
+
+        for (int i = startWord - 1; i < endWord; i++) {
```
Comment (jonathanjarman) [inline src/main/java/StringManipulation.java:83]: For this method rather than iterating through and copying one word at a time, you can use the arraycopy method to copy the array, which is more efficient. Picture for reference. 
image

File: src/main/java/org/example/Main.java [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package org.example;
```
Comment (jonathanjarman) [inline src/main/java/org/example/Main.java]: This entire file and folder should be deleted from this repo. It may be causing the build issue. 
Delete org/example - folder
Main.java  - file 

image

**Review by jonathanjarman — CHANGES_REQUESTED** [review]
This latest commit is very good. I would just try and add some comments to explain what the code is doing. I would also delete the extra Main.java file and org/example folder that are inside of this repo, as it may be why the build is not passing Travis CI. There were some extra local variables that were not being used in some of the test cases, which could be removed. Great job overall though, I noticed throughout the commit history there was steady improvement in the logic and design of the methods.

File: src/main/java/StringManipulation.java (hunk 13 above)
Comment (Hester1937) [reply src/main/java/StringManipulation.java:83]: thank you Jonathan

File: src/main/java/StringManipulation.java (hunk 6 above)
Comment (Hester1937) [reply src/main/java/StringManipulation.java:39]: I separate two if statements for null and isEmpty.

File: src/test/java/StringManipulationTest.java (hunk 8 above)
Comment (Hester1937) [reply src/test/java/StringManipulationTest.java]: I did change here

File: src/test/java/StringManipulationTest.java (hunk 9 above)
Comment (Hester1937) [reply src/test/java/StringManipulationTest.java]: I did change here

File: src/test/java/StringManipulationTest.java (hunk 10 above)
Comment (Hester1937) [reply src/test/java/StringManipulationTest.java]: Changed already

File: src/test/java/StringManipulationTest.java (hunk 11 above)
Comment (Hester1937) [reply src/test/java/StringManipulationTest.java]: changed already

File: src/test/java/StringManipulationTest.java (hunk 12 above)
Comment (Hester1937) [reply src/test/java/StringManipulationTest.java]: changed

File: src/main/java/org/example/Main.java (hunk 14 above)
Comment (Hester1937) [reply src/main/java/org/example/Main.java]: you are right. It is run successful after I delete the org/example

File: src/test/java/StringManipulationTest.java (hunk 7 above)
Comment (Hester1937) [reply src/test/java/StringManipulationTest.java]: I changed already
